llvm/lib/SYCL/KernelPropGen/KernelPropGen.cpp

```diff
@@ -0,0 +1,147 @@
+//===- KernelPropGen.cpp                                      ---------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// Retrieves the names of the kernels inside of the passed in file and places
+// them into a text file. Possible to merge this into another pass if
+// required, as it's a fairly trivial pass on its own.
+// ===---------------------------------------------------------------------===//
+
+#include <cstddef>
+#include <regex>
+#include <string>
+
+#include "llvm/SYCL/KernelPropGen.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+// Put the code in an anonymous namespace to avoid polluting the global
+// namespace
+namespace {
+
+/// Retrieve the names for all kernels in the module and place them into a file
+struct KernelPropGen : public ModulePass {
+
+  static char ID; // Pass identification, replacement for typeid
+
+  KernelPropGen() : ModulePass(ID) {}
+
+  bool isKernel(const Function &F) {
```

Comment:
It would be a useful addition long term although I don't think there is enough reusable functionality between all of the passes at the moment for it to be worth it, but I think as the passes grow that will change. But this might be coming back around to the idea of some kind of unified reusable library for the runtime and llvm passes i.e. https://github.com/triSYCL/sycl/issues/20